libfwupd/fwupd-client.c

```diff
@@ -121,13 +123,116 @@ fwupd_client_curl_helper_free (FwupdCurlHelper *helper)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC(FwupdCurlHelper, fwupd_client_curl_helper_free)
 #endif
 
+typedef struct {
+	FwupdClient	*self;
+	gchar		*property_name;
+	guint		 signal_id;
+	FwupdDevice	*device;
+} FwupdClientContextHelper;
+
+static void
+fwupd_client_context_helper_free (FwupdClientContextHelper *helper)
+{
+	if (helper->device != NULL)
+		g_object_unref (helper->device);
+	g_object_unref (helper->self);
+	g_free (helper->property_name);
+	g_free (helper);
+}
+
+static gboolean
+fwupd_client_context_idle_cb (gpointer user_data)
+{
+	FwupdClientContextHelper *helper = (FwupdClientContextHelper *) user_data;
+	FwupdClient *self = FWUPD_CLIENT (helper->self);
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GMutexLocker) locker = NULL;
+
+	/* property */
+	if (helper->property_name != NULL)
+		g_object_notify (G_OBJECT (self), helper->property_name);
+
+	/* legacy signal name */
+	if (g_strcmp0 (helper->property_name, "status") == 0)
+		g_signal_emit (self, signals[SIGNAL_STATUS_CHANGED], 0, priv->status);
+
+	/* device signal */
+	if (helper->property_name != NULL)
+		g_signal_emit (self, signals[helper->signal_id], 0, helper->device);
+
+	/* all done */
+	locker = g_mutex_locker_new (&priv->idle_id_mutex);
+	priv->idle_id = 0;
+	return G_SOURCE_REMOVE;
+}
+
+static void
+fwupd_client_context_helper (FwupdClient *self, FwupdClientContextHelper *helper)
+{
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&priv->idle_id_mutex);
+	g_autoptr(GSource) source = g_idle_source_new ();
+
+	/* run in the correct GMainContext and thread */
+	g_source_set_callback (source,
+			       fwupd_client_context_idle_cb,
+			       helper,
+			       (GDestroyNotify) fwupd_client_context_helper_free);
+	if (priv->idle_id != 0)
+		g_source_remove (priv->idle_id);
+	priv->idle_id = g_source_attach (g_steal_pointer (&source), priv->main_ctx);
+}
+
+/* run callback in the correct thread */
+static void
+fwupd_client_object_notify (FwupdClient *self, const gchar *property_name)
+{
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	FwupdClientContextHelper *helper = NULL;
+
+	/* shortcut */
+	if (priv->main_ctx == g_main_context_get_thread_default ()) {
+		g_object_notify (G_OBJECT (self), property_name);
+		if (g_strcmp0 (property_name, "status") == 0)
+			g_signal_emit (self, signals[SIGNAL_STATUS_CHANGED], 0, priv->status);
```

Comment:
Duplicating the code from `fwupd_client_context_idle_cb()` here means it’s possibly going to end up going out of sync.

Reply:
Maybe we add something like fwupd_client_set_status_safe() and do it there?

Reply:
This is what I did; better function name welcome.